Dockerfile

```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:18.04
+FROM nvidia/cuda:10.2-base-ubuntu18.04
```

Comment:
We have discussed this earlier. (Sorry, cannot find a link.) The problem is that the nvidia images are exclusive: you cannot run them without GPU hardware and nvidia driver on the host system. (Also, you cannot run them on a macOS or Windows host.) This disqualifies them for general-purpose deployments like OCR-D. What we need here is an _opt-in_. 

Maybe in the future we will address this by providing two separate variants of all images. But hopefully a better solution comes up.

Reply:
Thanks for the answer.

So for now, running torch/GPU based code is not possible from within the Docker images, correct? Could you please clarify this in the documentation and also list all affected parts of ocrd?

---

Creating separate image variants seems to be a matter of replacing

```
FROM ubuntu:18.04
```

with

```
FROM nvidia/cuda:10.2-base-ubuntu18.04
```

in `ocrc/core` and introducing new labels, such as `maximum-gpu`. Do you have an estimate when such images will become available? Anything I can do to support these changes?

---

Finally, I came across a recommendation about how to write device-agnostic code in pytorch, running both on CPU and GPU (if available).

This way, one could for example run `ocrd-anybaseocr-dewarp` without to go through the installation CUDA, or even on a system not having Nvidia GPU(s) at all.

See *Writing device-agnostic code* [here](https://github.com/pytorch/pytorch/releases/tag/v0.4.0).

Reply:
> So for now, running torch/GPU based code is not possible from within the Docker images, correct? 

If pytorch is not able to run without GPU, then yes. (Tensorflow always falls back to CPU.)

You'd need a local installation (preferably via OCR-D/ocrd_all) for GPU utilisation that ATM.

> Could you please clarify this in the documentation and also list all affected parts of ocrd?

Yes, we absolutely should. Generally, deployment options should be better documented IMO. Feel free to open issues to spec / core / ocrd_all. (But we will discuss this anyway, as soon as @kba and collegues are back from this week's DHd meeting.)

Reply:
> in `ocrd/core` and introducing new labels, such as `maximum-gpu`. Do you have an estimate when such images will become available? Anything I can do to support these changes?

Yes, providing `ocrd/core:gpu` and `ocrd/all:{mini,medi,maxi}mum-gpu` with a few changes to CircleCI/Docker(hub) configs would be a simple and sound provisional solution. This should not take more than a week to set up (but I can't tell what other commitments @kba has and whether he approves yet). I don't think there's anything you can do to speed this up right now.

Reply:
> Finally, I came across a recommendation about how to write device-agnostic code in pytorch, running both on CPU and GPU (if available).
> 
> This way, one could for example run `ocrd-anybaseocr-dewarp` without to go through the installation CUDA, or even on a system not having Nvidia GPU(s) at all.
> 
> See _Writing device-agnostic code_ [here](https://github.com/pytorch/pytorch/releases/tag/v0.4.0).
> 
> See also [this](https://pytorch.org/docs/stable/notes/cuda.html#device-agnostic-code).

Excellent! Then this recipe should be made a hard requirement for OCR-D modules using pytorch. In fact, `ocrd-typegroups-classifier` already implements this, cf [here](https://github.com/seuretm/ocrd_typegroups_classifier/blob/60293f87b79beb06a8e0cfc6824f83f8033c3475/ocrd_typegroups_classifier/typegroups_classifier.py#L52). 

Could you open issue/PR for `ocrd_anybaseocr`?

Reply:
> > Finally, I came across a recommendation about how to write device-agnostic code in pytorch, running both on CPU and GPU (if available).
> > This way, one could for example run `ocrd-anybaseocr-dewarp` without to go through the installation CUDA, or even on a system not having Nvidia GPU(s) at all.
> > See _Writing device-agnostic code_ [here](https://github.com/pytorch/pytorch/releases/tag/v0.4.0).
> > See also [this](https://pytorch.org/docs/stable/notes/cuda.html#device-agnostic-code).
> 
> Excellent! Then this recipe should be made a hard requirement for OCR-D modules using pytorch. In fact, `ocrd-typegroups-classifier` already implements this, cf [here](https://github.com/seuretm/ocrd_typegroups_classifier/blob/60293f87b79beb06a8e0cfc6824f83f8033c3475/ocrd_typegroups_classifier/typegroups_classifier.py#L52).
> 
> Could you open issue/PR for `ocrd_anybaseocr`?

I'm on it, see https://github.com/OCR-D/core/pull/452#issuecomment-596099737.